src/main/java/io/sqreen/powerwaf/PowerwafContext.java

```diff
@@ -152,10 +139,15 @@ public Powerwaf.ResultWithData runRules(Map<String, Object> persistentData,
                                 this);
                         throw new TimeoutPowerwafException();
                     }
-                    res = Powerwaf.runRules(
-                            this.handle, persistentBuffer, ephemeralBuffer, limits, metrics);
+                    if (isEphemeral) {
+                        res = Powerwaf.runRules(
+                                this.handle, null, lease.getFirstPWArgsByteBuffer(), newLimits, metrics);
+                    } else {
+                        res = Powerwaf.runRules(
+                                this.handle, lease.getFirstPWArgsByteBuffer(), null, newLimits, metrics);
+                    }
```

Comment:
You could collapse into the first branch (`isEphemeral`) because we're discarding the `ddwaf_context` after this call. This may improve the performance a bit because caches/saved matches are not created.

In fact, the native function Powerwaf#runRules doesn't need to take persistent and ephemeral data separately.

But like I said before, ultimately, in a later PR, the better option would be to completely remove this function (and `_run_rule_common` in the native code). If you need a one-shot ddwaf run, you can create an Additive and discard it afterwards; that will a bit of a penalty (more Java objects created), but at least for the Java tracer, there will always be an Additive associated with the request.

Reply:
Done